### PR TITLE
implement API for local-vars-hook

### PR DESCRIPTION
**Pull request by d12frosted**
Following https://github.com/syl20bnr/spacemacs/issues/6419#issuecomment-228298464 and https://github.com/syl20bnr/spacemacs/issues/6419#issuecomment-228327497.

### CODE REVIEW

**Comment by d12frosted** [comment]
I've also pushed to this branch haskell layer update to support this flow and fix for ruby layer.

@TheBB @syl20bnr I can separate it to tree distinct PRs if you think that it would be better

File: layers/+distributions/spacemacs-base/config.el [hunk 1]
```diff
@@ -189,3 +189,11 @@ It runs `tabulated-list-revert-hook', then calls `tabulated-list-print'."
...
   (remove-hook 'kill-buffer-query-functions 'server-kill-buffer-query-function))
 (add-hook 'server-visit-hook 'server-remove-kill-buffer-hook)
+
+;; ---------------------------------------------------------------------------
+;; Other
+;; ---------------------------------------------------------------------------
+
+;; hook into `hack-local-variables' in order to allow switching spacemacs
+;; configurations based on local variables
+(add-hook #'hack-local-variables-hook #'spacemacs//run-local-vars-mode-hook)
```
Comment (TheBB) [inline layers/+distributions/spacemacs-base/config.el]: Should probably not be `#` on the first one.

File: layers/+lang/elixir/packages.el [hunk 2]
```diff
@@ -124,8 +124,8 @@
                    (cons 'elixir-enable-compilation-checking nil))
       (add-to-list 'safe-local-variable-values
                    (cons 'elixir-enable-compilation-checking t))
-      (add-hook 'elixir-mode-hook
-                'spacemacs//elixir-enable-compilation-checking t))))
+      (spacemacs/add-to-local-vars-hook 'elixir-mode-hook
```
Comment (TheBB) [inline layers/+lang/elixir/packages.el]: Judging from the two functions in spacemacs-base/funcs.el you have to use `elixir-mode` here?

File: layers/+distributions/spacemacs-base/config.el (hunk 1 above)
Comment (d12frosted) [inline layers/+distributions/spacemacs-base/config.el]: Indeed, it's var, not func. 👍

File: layers/+lang/elixir/packages.el (hunk 2 above)
Comment (d12frosted) [inline layers/+lang/elixir/packages.el]: Thanks for catching this.

**Comment by d12frosted** [comment]
Also I think that this feature should be documented. General note in documentation and links from layers that support such configuration.

**Comment by syl20bnr** [comment]
Excellent 💯 
I removed the helper because it was abstracting a one liner so we are better to use the standard functions to hook functions.
Also instead of `(concat ....)` with symbol names you can use `(format "%S" ...)`
Thank you ! 👍 
Cherry-picked into develop branch, you can safely delete your branch.

**Comment by d12frosted** [comment]
> I removed the helper because it was abstracting a one liner so we are better to use the standard functions to hook functions.

Makes sense. And looks much shorter. Thanks.

As per `concat` vs `format` - didn't know about `%S`. Thanks. 

P. S. commit for reference: https://github.com/syl20bnr/spacemacs/commit/89dc1af9a60e5d8bdf58e345c46b78df39835fbd

**Pull request closed without merging**